src/compiler/checker.ts

```diff
@@ -24062,12 +24097,61 @@ namespace ts {
                 return getApplicableIndexInfoForName(type, propName) ? true : !assumeTrue;
             }
 
-            function narrowByInKeyword(type: Type, name: __String, assumeTrue: boolean) {
-                if (type.flags & TypeFlags.Union
-                    || type.flags & TypeFlags.Object && declaredType !== type
-                    || isThisTypeParameter(type)
-                    || type.flags & TypeFlags.Intersection && every((type as IntersectionType).types, t => t.symbol !== globalThisSymbol)) {
-                    return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+            function widenTypeWithSymbol(type: Type, newSymbol: Symbol): Type {
+                // If type is this/any/unknown, it cannot be widened.
+                if ((type.flags & TypeFlags.AnyOrUnknown) || isThisTypeParameter(type)) {
+                    return type;
+                }
+                // If type is anonymous object, add the symbol directly
+                if (isObjectType(type) && type.objectFlags & ObjectFlags.Anonymous) {
+                    return widenObjectType(type, newSymbol);
+                }
+                // If type is intersection, add the symbol to the first anonymous object component of the intersection
+                if (isIntersectionType(type)) {
+                    const objectSubtype = type.types.find(t => isObjectType(t) && t.objectFlags & ObjectFlags.Anonymous) as ObjectType | undefined;
+                    if (objectSubtype) {
+                        const restOfIntersection = filterUnionOrIntersectionType(type, t => t !== objectSubtype);
+                        return createIntersectionType([restOfIntersection, widenObjectType(objectSubtype, newSymbol)]);
+                    }
+                }
+
+                // Otherwise, just add the new object type as an intersection
+                const newTypeWithSymbol = widenObjectType(createAnonymousType(undefined, createSymbolTable(), emptyArray, emptyArray, emptyArray), newSymbol);
+                return createIntersectionType([type, newTypeWithSymbol]);
+
+                function widenObjectType(type: ObjectType, newSymbol: Symbol): Type {
+                    const members = createSymbolTable();
+                    if (type.members !== undefined) {
+                        mergeSymbolTable(members, type.members);

```

Comment:
> Ideally, I think I would like to clone the symbol table and just add the new symbol to it (that way there are no merges), but I can't figure out how to clone a symbol table :/

Don't merge the types :)

Instead, just make an anonymous type with a single member - the new one - and use `getSpreadType` to combine the original type with that new one. It'll combine directly where it can, and intersect otherwise.

Reply:
Essentially everything you're doing by hand here in the last 3/4ths of `widenTypeWithSymbol`, but already done for you. 😊

Reply:
Hi,
I tried to do it this way:
```diff
-                    return widenTypeWithSymbol(type, addSymbol);
+                    const members = createSymbolTable([addSymbol]);
+                    const addType = createAnonymousType(undefined, members, emptyArray, emptyArray, emptyArray);
+                    return getSpreadType(type, addType, undefined, getObjectFlags(type), false);
```
It's a really neat solution that removes the entire `widenTypeWithSymbol` function, thanks. However, it discards all call signatures of `type`! Looking at the `getSpreadType` source, I could probably modify that to keep them, but I'm not sure that wouldn't break other stuff.

See it in action [here](https://github.com/marekdedic/TypeScript/tree/in-operator-add-property-getSpreadType)